crates/services/gas_price_service/src/v1/da_source_adapter/block_committer_costs.rs

```diff
@@ -6,57 +8,321 @@ use crate::v1::da_source_adapter::{
+            Ok(self.value.clone())
+        }
+        async fn get_costs_by_seqno(
+            &self,
+            seq_no: u64,
+        ) -> DaBlockCostsResult<Option<RawDaBlockCosts>> {
+            // arbitrary logic to generate a new value
+            let mut value = self.value.clone();
+            if let Some(value) = &mut value {
+                value.sequence_number = seq_no;
+                value.blocks_range =
+                    value.blocks_range.end * seq_no..value.blocks_range.end * seq_no + 10;
+                value.da_block_height = value.da_block_height + (seq_no + 1).into();
+                value.total_cost += 1;
+                value.total_size_bytes += 1;
+            }
+            Ok(value)
+        }
+        async fn get_cost_bundles_by_range(
+            &self,
+            _: core::ops::Range<u64>,
+        ) -> DaBlockCostsResult<Vec<Option<RawDaBlockCosts>>> {
+            Ok(vec![self.value.clone()])
+        }
+    }
+
+    fn test_da_block_costs() -> RawDaBlockCosts {
+        RawDaBlockCosts {
+            sequence_number: 1,
+            blocks_range: 0..10,
+            da_block_height: 1u64.into(),
+            total_cost: 1,
+            total_size_bytes: 1,
+        }
+    }
+
+    #[tokio::test]
+    async fn request_da_block_cost__when_last_value_is_none__then_get_latest_costs_is_called(
+    ) {
+        // given
+        let da_block_costs = test_da_block_costs();
+        let expected = (&da_block_costs).into();
+        let mock_api = MockBlockCommitterApi::new(Some(da_block_costs));
+        let mut block_committer = BlockCommitterDaBlockCosts::new(mock_api, None);
+
+        // when
+        let actual = block_committer.request_da_block_cost().await.unwrap();
+
+        // then
+        assert_eq!(actual, expected);
+        assert!(block_committer.last_raw_da_block_costs.is_some());
+    }
+
+    #[tokio::test]
+    async fn request_da_block_cost__when_last_value_is_some__then_get_costs_by_seqno_is_called(
+    ) {
+        // given
+        let mut da_block_costs = test_da_block_costs();
+        let mock_api = MockBlockCommitterApi::new(Some(da_block_costs.clone()));
+        let mut block_committer =
+            BlockCommitterDaBlockCosts::new(mock_api, Some(da_block_costs.clone()));
+
+        // when
+        let actual = block_committer.request_da_block_cost().await.unwrap();
+
+        // then
+        assert_ne!(da_block_costs.blocks_range, actual.l2_block_range);
+    }
+
+    #[tokio::test]
+    async fn request_da_block_cost__when_response_is_none__then_error() {
+        // given
+        let mock_api = MockBlockCommitterApi::new(None);
+        let mut block_committer = BlockCommitterDaBlockCosts::new(mock_api, None);
+
+        // when
+        let result = block_committer.request_da_block_cost().await;
+
+        // then
+        assert!(result.is_err());
+        assert!(block_committer.last_raw_da_block_costs.is_none());
```

Comment:
I think this check is unnecessary. From the outside observer, we don't care what is being set internally, just the external result of those internal values.

Reply:
addressed in 03d39c7ac6